# Data Mining 2024-1 과제 \#1

## 1. 목표 <a id="intro"></a>

1. 해당 jupyter notebook 내의 지시사항을 따라 미완성된 코드를 모두 작성하세요.

2. 아래의 분류 모델을 사용하여 주어진 데이터셋에 대한 정확도(Accuracy)를 구하는 코드를 완성해야 합니다.

    - Decision Tree
    - Naive Bayes
    - Random Forest
    - Gradient Boost (LightGBM, XGBoost)
    - Support Vector Machine (SVM)
    - Multi Layer Perceptron (MLP)
    - Bagging
    - Boosting
    - Voting (Random Forest, XGBoost, LightGBM, SVM, MLP)
    - Your own method

3. 자신만의 방법으로 구현 시, 본인이 알고 있는 여러 방법을 적용하여 가장 높은 Accuracy를 보일 수 있는 코드를 제출해야 합니다.

## 2. 환경 설정

아래의 코드는 Python >= 3.6 를 만족하는 환경과 호환됩니다.

채점 시 오류가 발생하지 않도록 본인이 작성한 코드 또한 호환되어야 합니다.

Virtual Environment를 사용하여 위 version을 만족하는 환경을 구축하는 것을 권장합니다.

아래 명시된 package 외에 추가적인 package를 설치하여 사용하는 것은 금합니다.


1. **[scikit-learn](https://scikit-learn.org/stable/index.html)**

2. **[numpy](https://numpy.org/doc/stable/)**

3. **[pandas](https://pandas.pydata.org/docs/)**

4. **[lightgbm](https://lightgbm.readthedocs.io/en/stable/)**

5. **[xgboost](https://xgboost.readthedocs.io/en/stable/)**

과제를 수행하기 전 명시된 package를 모두 설치하세요.

주어진 package는 되도록 최신 version을 설치하세요.

## 3. 구현사항 - 데이터셋

### 3-1 데이터셋 소개

본 과제에서는 다음의 데이터셋을 사용할 것입니다: `Pima Indian Diabetes Dataset`

* Pima Indian Diabetes Dataset

    이 데이터셋에서는 768명의 여성들에 대한 8 가지의 의료 수치 항목을 바탕으로 당뇨병 여부를 예측해야 합니다. (Binary Classification)

해당 데이터셋은 `csv` 포맷의 파일로 주어졌습니다.

이를 `pandas`의 `DataFrame`으로 불러와 사용해야 합니다.

### 3-2. Load Pima Indian Diabetes Dataset **(주석 내에 코드를 작성하세요)**

In [5]:
import pandas as pd
import numpy as np
import os

DATA_DIR = "/content"   #"./data"
PID_CSV_FILENAME = "diabetes.csv" # PID(Pima Indian Diabetes)

############### Put your code here ###############
#sample_data 옆에 저장함.
#엑셀에서 =COUNTIF() 를 이용한 사전 처리 과정에서 768개의 데이터중 Outcome에는 0은 500개, 1은 268개 있음.
file_path = f"{DATA_DIR}/{PID_CSV_FILENAME}"
df = pd.read_csv(file_path)
print(df)
##################################################

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                  

### 3-4. Pima Indian Diabetesa Dataset Preprocessing **(주석 내에 코드를 작성하세요)**

이 과정에서는 데이터셋을 분석하여 필요하다면 적절히 가공하고,

학습 및 평가를 진행하기 전, 데이터셋을 `train`과 `test` subset으로 나눌 것입니다.

더 높은 정확도를 위해 필요하다면 본인만의 방법을 적용하여도 좋습니다.

In [6]:
from sklearn.model_selection import train_test_split

############### Put your code here ###############
#사전처리 결과에 따르면 0과 1의 비율이 대략 500 : 268 이므로 stratify옵션을 통해 원본의 대략 65%, 35% 비율과 근사한 데이터 집합을 만들겠다.
X = df[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction','Age' ]]
y = df['Outcome']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, train_size = 0.75, random_state = 1, shuffle = True, stratify = y)
print(y)
##################################################

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


### 4. 구현사항 - 분류모델 학습과 평가 **(주석 내에 코드를 작성하세요)**

[목표](#Intro)에서 설명하였듯이, 본인만의 방법을 포함하여 총 11가지의 분류 모델을 사용합니다.

아래 목록과 코드를 참고하여 분류모델을 `train` subset으로 학습하고 `test` subset에 대한 정확도를 출력하는 코드를 작성하세요.

Voting Classifier는 앞서 선언한 `Random Forest`, `LightGBM`, `XGBoost`, `SVM`, `MLP` 5가지의 분류 모델을 포함해야 합니다.

- Decision Tree
- Naive Bayes (Gaussian)
- Random Forest
- LightGBM
- XGBoost
- Support Vector Machine (SVM)
- Multi Layer Perceptron (MLP)
- Bagging
- Boosting (AdaBoost)
- Voting (Random Forest, XGBoost, LightGBM, SVM, MLP)
- Your own method

### 4-1. Decision Tree

scikit-learn에서 제공하는 `Decision Tree Classifier`를 `Pima Indian Diabetes Dataset`의 `train` subset으로 학습시키고 `test` subset에 대한 정확도를 출력하세요.

In [7]:
from sklearn.tree import DecisionTreeClassifier

############### Put your code here ###############
for i in range(1,10):
    model = DecisionTreeClassifier(max_depth= i, random_state = 1).fit(X_train,y_train)
    print("깊이=",i,"에 대한 train 정확도는", model.score(X_train,y_train), "test 정확도는", model.score(X_test,y_test))

#깊이=6 이상에서 overfitting이 나타나므로 최적의 깊이는 5이다.
##################################################

깊이= 1 에 대한 train 정확도는 0.7430555555555556 test 정확도는 0.7135416666666666
깊이= 2 에 대한 train 정확도는 0.78125 test 정확도는 0.7447916666666666
깊이= 3 에 대한 train 정확도는 0.7847222222222222 test 정확도는 0.7604166666666666
깊이= 4 에 대한 train 정확도는 0.8003472222222222 test 정확도는 0.7552083333333334
깊이= 5 에 대한 train 정확도는 0.8420138888888888 test 정확도는 0.7760416666666666
깊이= 6 에 대한 train 정확도는 0.8715277777777778 test 정확도는 0.7447916666666666
깊이= 7 에 대한 train 정확도는 0.9097222222222222 test 정확도는 0.734375
깊이= 8 에 대한 train 정확도는 0.9340277777777778 test 정확도는 0.7552083333333334
깊이= 9 에 대한 train 정확도는 0.9479166666666666 test 정확도는 0.7395833333333334


### 4-2. Naive Bayes

scikit-learn에서 제공하는 `Gausian Naive Bayes`를 `Pima Indian Diabetes Dataset`의 `train` subset으로 학습시키고 `test` subset에 대한 정확도를 출력하세요.

In [8]:
from sklearn.naive_bayes import GaussianNB

############### Put your code here ###############
from sklearn.metrics import accuracy_score
model = GaussianNB()
model.fit(X_train,y_train)
predicted = model.predict(X_test)
print("test 정확도는",accuracy_score(y_test, predicted))
##################################################

test 정확도는 0.7135416666666666


### 4-3. Random Forest

scikit-learn에서 제공하는 `Random Forest Classifier`를 `Pima Indian Diabetes Dataset`의 `train` subset으로 학습시키고 `test` subset에 대한 정확도를 출력하세요.

In [9]:
from sklearn.ensemble import RandomForestClassifier

############### Put your code here ###############
model = RandomForestClassifier(n_estimators=5, random_state=0)
model.fit(X_train, y_train)
print("test 정확도는", model.score(X_test,y_test))
##################################################

test 정확도는 0.734375


### 4-4. LightGBM

lightgbm에서 제공하는 `LGBMClassifier`를 `Pima Indian Diabetes Dataset`의 `train` subset으로 학습시키고 `test` subset에 대한 정확도를 출력하세요.

In [17]:
from lightgbm import LGBMClassifier

############### Put your code here ###############
lgbm = LGBMClassifier()
lgbm.fit(X_train, y_train)

# 테스트 데이터로 예측
y_pred = lgbm.predict(X_test)

# 정확도 출력
accuracy = accuracy_score(y_test, y_pred)
print("test 정확도는:", accuracy)
##################################################

[LightGBM] [Info] Number of positive: 201, number of negative: 375
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000118 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 642
[LightGBM] [Info] Number of data points in the train set: 576, number of used features: 8
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.348958 -> initscore=-0.623621
[LightGBM] [Info] Start training from score -0.623621
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best 

### 4-5. XGBoost

xgboost에서 제공하는 `XGBClassifier`를 `Pima Indian Diabetes Dataset`의 `train` subset으로 학습시키고 `test` subset에 대한 정확도를 출력하세요.

In [11]:
from xgboost import XGBClassifier

############### Put your code here ###############
model = XGBClassifier()
model.fit(X_train, y_train)
print("test 정확도는", model.score(X_test,y_test))
##################################################

test 정확도는 0.7291666666666666


### 4-6. SVM(Support Vector Machine)

scikit-learn에서 제공하는 `SVC`를 `Pima Indian Diabetes Dataset`의 `train` subset으로 학습시키고 `test` subset에 대한 정확도를 출력하세요.

In [12]:
from sklearn.svm import SVC

############### Put your code here ###############
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
clf = make_pipeline(StandardScaler(), SVC(gamma='auto'))
clf.fit(X_train, y_train)

print("test 정확도는", clf.score(X_test,y_test))
##################################################

test 정확도는 0.7135416666666666


### 4-7. MLP(Multi-Layer Perceptron)

scikit-learn에서 제공하는 `MLPClassifier`를 `Pima Indian Diabetes Dataset`의 `train` subset으로 학습시키고 `test` subset에 대한 정확도를 출력하세요.

In [13]:
from sklearn.neural_network import MLPClassifier

############### Put your code here ###############
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X_train)
X_train1 = scaler.transform(X_train)
X_test1 = scaler.transform(X_test)

mlp = MLPClassifier(hidden_layer_sizes=(10,10,10), max_iter=100)
mlp.fit(X_train1, y_train)
print("test 정확도는", mlp.score(X_test1,y_test))
##################################################

test 정확도는 0.7083333333333334


/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(


### 4-8. Bagging

scikit-learn에서 제공하는 `Bagging Classifier`를 `Pima Indian Diabetes Dataset`의 `train` subset으로 학습시키고 `test` subset에 대한 정확도를 출력하세요.

In [14]:
from sklearn.ensemble import BaggingClassifier

############### Put your code here ###############
clf = BaggingClassifier(estimator=SVC(),n_estimators=10, random_state=0).fit(X_train, y_train)
print("test 정확도는", clf.score(X_test,y_test))
##################################################

test 정확도는 0.7239583333333334


### 4-9. Boosting

scikit-learn에서 제공하는 `AdaBoost Classifier`를 `Pima Indian Diabetes Dataset`의 `train` subset으로 학습시키고 `test` subset에 대한 정확도를 출력하세요.

In [15]:
from sklearn.ensemble import AdaBoostClassifier

############### Put your code here ###############
clf = AdaBoostClassifier(n_estimators=100, algorithm="SAMME", random_state=0)
clf.fit(X_train, y_train)
print("test 정확도는", clf.score(X_test,y_test))
##################################################

test 정확도는 0.75


### 4-10. Voting

scikit-learn에서 제공하는 `Voting Classifier`를 `Pima Indian Diabetes Dataset`의 `train` subset으로 학습시키고 `test` subset에 대한 정확도를 출력하세요.

Voting Classifier는 `Random Forest`, `LightGBM`, `XGBoost`, `SVM`, `MLP` 총 5 가지의 분류 모델을 포함해야 합니다.

In [18]:
from sklearn.ensemble import VotingClassifier

############### Put your code here ###############
# 분류기 정의
rf_clf = RandomForestClassifier()
lgbm_clf = LGBMClassifier()
xgb_clf = XGBClassifier()
svm_clf = SVC(probability=True)  # Voting Classifier에서 확률을 사용하기 위해 probability=True 설정
mlp_clf = MLPClassifier()

# Voting Classifier 생성
voting_clf = VotingClassifier(estimators=[
    ('rf', rf_clf),
    ('lgbm', lgbm_clf),
    ('xgb', xgb_clf),
    ('svm', svm_clf),
    ('mlp', mlp_clf)
], voting='hard')  # 'hard' voting은 각 분류기의 결과를 다수결로 결정

# Voting Classifier 학습
voting_clf.fit(X_train, y_train)

# 테스트 데이터로 예측
y_pred = voting_clf.predict(X_test)

# 정확도 출력
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)
##################################################

[LightGBM] [Info] Number of positive: 201, number of negative: 375
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000121 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 642
[LightGBM] [Info] Number of data points in the train set: 576, number of used features: 8
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.348958 -> initscore=-0.623621
[LightGBM] [Info] Start training from score -0.623621
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best 

### 4-11. Your own method

자신만의 방법으로 구현된 분류 모델을 `Pima Indian Diabetes Dataset`의 `train` subset으로 학습시키고 `test` subset에 대한 정확도를 출력하세요.

필요하다면 주어진 Dataset을 추가적으로 가공하여 가장 높은 정확도를 달성하는 코드를 작성하세요.

 **주어진 데이터셋에 가장 높은 정확도를 보인 model은 DecisionTreeClassifier였다.따라서 해당 model의 파라미터를 바꾸어가며 최적화시키도록 하겠다. 우선 가장 중요한 트리의 최적 max_depth 후보를 선별하겠다.**

In [26]:
for i in range(1,20):
    model = DecisionTreeClassifier(max_depth= i, random_state = 1).fit(X_train,y_train)
    print("깊이=",i,"에 대한 train 정확도는", model.score(X_train,y_train), "test 정확도는", model.score(X_test,y_test))

깊이= 1 에 대한 train 정확도는 0.7430555555555556 test 정확도는 0.7135416666666666
깊이= 2 에 대한 train 정확도는 0.78125 test 정확도는 0.7447916666666666
깊이= 3 에 대한 train 정확도는 0.7847222222222222 test 정확도는 0.7604166666666666
깊이= 4 에 대한 train 정확도는 0.8003472222222222 test 정확도는 0.7552083333333334
깊이= 5 에 대한 train 정확도는 0.8420138888888888 test 정확도는 0.7760416666666666
깊이= 6 에 대한 train 정확도는 0.8715277777777778 test 정확도는 0.7447916666666666
깊이= 7 에 대한 train 정확도는 0.9097222222222222 test 정확도는 0.734375
깊이= 8 에 대한 train 정확도는 0.9340277777777778 test 정확도는 0.7552083333333334
깊이= 9 에 대한 train 정확도는 0.9479166666666666 test 정확도는 0.7395833333333334
깊이= 10 에 대한 train 정확도는 0.96875 test 정확도는 0.7239583333333334
깊이= 11 에 대한 train 정확도는 0.984375 test 정확도는 0.7239583333333334
깊이= 12 에 대한 train 정확도는 0.9947916666666666 test 정확도는 0.7291666666666666
깊이= 13 에 대한 train 정확도는 1.0 test 정확도는 0.7239583333333334
깊이= 14 에 대한 train 정확도는 1.0 test 정확도는 0.7239583333333334
깊이= 15 에 대한 train 정확도는 1.0 test 정확도는 0.7239583333333334
깊이= 16 에 대한 train 정확도는 1.0 test

**깊이는 iteration을 반복할 수록 0.72395에 수렴해가므로 이 보다 큰 정확도를 갖는 시도들 중 max_depth= 3,4,5,8 만 선별하여 random_state를 바꾸어가며 평균 성능을 구하고, max_depth를 하나로 정하겠다.**

In [35]:
#max_depth = 3일때
test_accuracy = []
for i in range(1,10):
    model = DecisionTreeClassifier(max_depth= 3, random_state = i).fit(X_train,y_train)
    print("random_state=",i,"에 대한 train 정확도는", model.score(X_train,y_train), "test 정확도는", model.score(X_test,y_test))
    test_accuracy.append(model.score(X_test, y_test))
print("평균 정확도:",np.mean(test_accuracy))

random_state= 1 에 대한 train 정확도는 0.7847222222222222 test 정확도는 0.7604166666666666
random_state= 2 에 대한 train 정확도는 0.7847222222222222 test 정확도는 0.7604166666666666
random_state= 3 에 대한 train 정확도는 0.7847222222222222 test 정확도는 0.7604166666666666
random_state= 4 에 대한 train 정확도는 0.7847222222222222 test 정확도는 0.7604166666666666
random_state= 5 에 대한 train 정확도는 0.7847222222222222 test 정확도는 0.7604166666666666
random_state= 6 에 대한 train 정확도는 0.7847222222222222 test 정확도는 0.7604166666666666
random_state= 7 에 대한 train 정확도는 0.7847222222222222 test 정확도는 0.7604166666666666
random_state= 8 에 대한 train 정확도는 0.7847222222222222 test 정확도는 0.7604166666666666
random_state= 9 에 대한 train 정확도는 0.7847222222222222 test 정확도는 0.7604166666666666
평균 정확도: 0.7604166666666666


In [36]:
#max_depth = 4일때
test_accuracy = []
for i in range(1,10):
    model = DecisionTreeClassifier(max_depth= 4, random_state = i).fit(X_train,y_train)
    print("random_state=",i,"에 대한 train 정확도는", model.score(X_train,y_train), "test 정확도는", model.score(X_test,y_test))
    test_accuracy.append(model.score(X_test, y_test))
print("평균 정확도:",np.mean(test_accuracy))

random_state= 1 에 대한 train 정확도는 0.8003472222222222 test 정확도는 0.7552083333333334
random_state= 2 에 대한 train 정확도는 0.8003472222222222 test 정확도는 0.7552083333333334
random_state= 3 에 대한 train 정확도는 0.8003472222222222 test 정확도는 0.7552083333333334
random_state= 4 에 대한 train 정확도는 0.8003472222222222 test 정확도는 0.7552083333333334
random_state= 5 에 대한 train 정확도는 0.8003472222222222 test 정확도는 0.7552083333333334
random_state= 6 에 대한 train 정확도는 0.8003472222222222 test 정확도는 0.7552083333333334
random_state= 7 에 대한 train 정확도는 0.8003472222222222 test 정확도는 0.7552083333333334
random_state= 8 에 대한 train 정확도는 0.8003472222222222 test 정확도는 0.7552083333333334
random_state= 9 에 대한 train 정확도는 0.8003472222222222 test 정확도는 0.7552083333333334
평균 정확도: 0.7552083333333334


In [37]:
#max_depth = 5일때
test_accuracy = []
for i in range(1,10):
    model = DecisionTreeClassifier(max_depth= 5, random_state = i).fit(X_train,y_train)
    print("random_state=",i,"에 대한 train 정확도는", model.score(X_train,y_train), "test 정확도는", model.score(X_test,y_test))
    test_accuracy.append(model.score(X_test, y_test))
print("평균 정확도:",np.mean(test_accuracy))

random_state= 1 에 대한 train 정확도는 0.8420138888888888 test 정확도는 0.7760416666666666
random_state= 2 에 대한 train 정확도는 0.8420138888888888 test 정확도는 0.7760416666666666
random_state= 3 에 대한 train 정확도는 0.8420138888888888 test 정확도는 0.7604166666666666
random_state= 4 에 대한 train 정확도는 0.8420138888888888 test 정확도는 0.7604166666666666
random_state= 5 에 대한 train 정확도는 0.8420138888888888 test 정확도는 0.7760416666666666
random_state= 6 에 대한 train 정확도는 0.8420138888888888 test 정확도는 0.7760416666666666
random_state= 7 에 대한 train 정확도는 0.8420138888888888 test 정확도는 0.7760416666666666
random_state= 8 에 대한 train 정확도는 0.8420138888888888 test 정확도는 0.7760416666666666
random_state= 9 에 대한 train 정확도는 0.8420138888888888 test 정확도는 0.7604166666666666
평균 정확도: 0.7708333333333334


In [38]:
#max_depth = 8일때
test_accuracy = []
for i in range(1,10):
    model = DecisionTreeClassifier(max_depth= 8, random_state = i).fit(X_train,y_train)
    print("random_state=",i,"에 대한 train 정확도는", model.score(X_train,y_train), "test 정확도는", model.score(X_test,y_test))
    test_accuracy.append(model.score(X_test, y_test))
print("평균 정확도:",np.mean(test_accuracy))

random_state= 1 에 대한 train 정확도는 0.9340277777777778 test 정확도는 0.7552083333333334
random_state= 2 에 대한 train 정확도는 0.9340277777777778 test 정확도는 0.734375
random_state= 3 에 대한 train 정확도는 0.9340277777777778 test 정확도는 0.7552083333333334
random_state= 4 에 대한 train 정확도는 0.9340277777777778 test 정확도는 0.7395833333333334
random_state= 5 에 대한 train 정확도는 0.9340277777777778 test 정확도는 0.7447916666666666
random_state= 6 에 대한 train 정확도는 0.9340277777777778 test 정확도는 0.734375
random_state= 7 에 대한 train 정확도는 0.9340277777777778 test 정확도는 0.7395833333333334
random_state= 8 에 대한 train 정확도는 0.9340277777777778 test 정확도는 0.7552083333333334
random_state= 9 에 대한 train 정확도는 0.9340277777777778 test 정확도는 0.7447916666666666
평균 정확도: 0.7447916666666666


**따라서 최적의 max_depth는 5이며, random_state 10개로 측정한 분류의 평균 정확도는  0.7708333333333334 이다.**